mongo/integration/sessions_test.go

```diff
@@ -358,6 +360,125 @@ func TestSessions(t *testing.T) {
+				name := "findOneAndDelete"
+				if err := result.Err(); err != nil && err != mongo.ErrNoDocuments {
+					return name, err
+				}
+				return name, nil
+			},
+			func(ctx context.Context) (string, error) {
+				result := mt.Coll.FindOneAndUpdate(ctx, bson.D{},
+					bson.D{{"$set", bson.D{{"a", 1}}}})
+
+				name := "findOneAndUpdate"
+				if err := result.Err(); err != nil && err != mongo.ErrNoDocuments {
+					return name, err
+				}
+				return name, nil
+			},
+			func(ctx context.Context) (string, error) {
+				result := mt.Coll.FindOneAndReplace(ctx, bson.D{}, bson.D{{"a", 1}})
+				name := "findOneAndReplace"
+				if err := result.Err(); err != nil && err != mongo.ErrNoDocuments {
+					return name, err
+				}
+				return name, nil
+			},
+			func(ctx context.Context) (string, error) {
+				name := "find"
+				cursor, err := mt.Coll.Find(ctx, bson.D{})
+				if err != nil {
+					return name, err
+				}
+				return name, cursor.All(ctx, &bson.A{})
+			},
+		}
+
+		// maintainedOneSession asserts that exactly one session is used for all operations at least once
+		// across the retries of this test.
+		var maintainedOneSession bool
+
+		// minimumSessionCount asserts the least amount of sessions used over all the retries of the
+		// operations. For example, if we retry 5 times we could result in session use { 1, 2, 1, 1, 6 }. In
+		// this case, minimumSessionCount should be 1.
+		var minimumSessionCount int
+
+		// limitedSessionUse asserts that the number of allocated sessions is strictly less than the number of
+		// concurrent operations in every retry of this test. In this instance it would be less than (but NOT
+		// equal to the number of operations).
+		limitedSessionUse := true
+
+		retrycount := 5
+		for i := 1; i <= retrycount; i++ {
+			errs, ctx := errgroup.WithContext(context.Background())
+
+			// Execute the ops list concurrently.
+			for _, op := range ops {
+				op := op
+				errs.Go(func() error {
+					cmd, err := op(ctx)
+					if err != nil {
+						return fmt.Errorf("error running %s operation: %v", cmd, err)
+					}
+					return nil
+				})
+			}
+			err := errs.Wait()
+			assert.Nil(mt, err, "expected no error, go: %v", err)
+
+			// Get all started events and collect them by the session ID.
+			set := make(map[string]bool)
+			for _, event := range mt.GetAllStartedEvents() {
+				lsid := event.Command.Lookup("lsid")
+				set[lsid.String()] = true
+			}
+
+			setSize := len(set)
+			if setSize == 1 {
+				maintainedOneSession = true
+			} else if setSize < minimumSessionCount || minimumSessionCount == 0 {
+				// record the minimum number of sessions we used over all retries.
+				minimumSessionCount = setSize
+			}
+
+			if setSize >= len(ops) {
+				limitedSessionUse = false
+			}
+		}
+
+		oneSessMsg := "expected one session across all %v operations for at least 1/%v retries, got: %v"
+		assert.True(mt, maintainedOneSession, oneSessMsg, len(ops), retrycount, minimumSessionCount)
```

Comment:
The `maintainedOneSession` assertion variable can be set to `true` on one loop and not reset back to `false` if a subsequent loop doesn't satisfy the `setSize == 1` condition. My understanding of the prose test is that we should make this assertion for every retry of the test logic. Move this assertion logic into the "retries" loop so we're asserting it on every retry.

E.g.
```go
// ...
setSize := len(set)
assert.True(mt,
	setSize == 1,
	"expected one session across all %v operations for at least 1/%v retries, got: %v",
	len(ops),
	retryCount,
	setSize)
// ...
```

Reply:
The phrasing for [this](https://github.com/mongodb/specifications/blob/master/source/sessions/tests/README.rst#implicit-sessions-only-allocate-their-server-session-after-a-successful-connection-checkout) is kind of strange, here is the assertion:

> Assert the following across at least 5 retries of the above test:
Drivers MUST assert that exactly one session is used for all operations at least once across the retries of this test.

I interpret this to mean that of the five sessions that run, only one has to maintain only one session.

Reply:
@prestonvasquez that is my interpretation as well and I believe your current assertion is correct.

Reply:
Agreed, I reached the same conclusion after re-reading it.